forge/src/trace/decoder.rs

```diff
@@ -0,0 +1,234 @@
+                                        );
+                                    }
+                                } else if let Ok(decoded_error) =
+                                    foundry_utils::decode_revert(&bytes[..], Some(&self.errors))
+                                {
+                                    node.trace.output = RawOrDecodedReturnData::Decoded(format!(
+                                        r#""{}""#,
+                                        decoded_error
+                                    ));
+                                }
+                            }
+                        }
+                    }
+                } else {
+                    node.trace.data = RawOrDecodedCall::Decoded("fallback".to_string(), Vec::new());
+
+                    if let RawOrDecodedReturnData::Raw(bytes) = &node.trace.output {
+                        if !node.trace.success {
+                            if let Ok(decoded_error) =
+                                foundry_utils::decode_revert(&bytes[..], Some(&self.errors))
+                            {
+                                node.trace.output = RawOrDecodedReturnData::Decoded(format!(
+                                    r#""{}""#,
+                                    decoded_error
+                                ));
+                            }
+                        }
+                    }
+                }
+            }
+
+            // Decode events
+            node.logs.iter_mut().for_each(|log| {
+                if let RawOrDecodedLog::Raw(raw_log) = log {
+                    if let Some(event) = self.events.get(&raw_log.topics[0]) {
+                        if let Ok(decoded) = event.parse_log(raw_log.clone()) {
+                            *log = RawOrDecodedLog::Decoded(
+                                event.name.clone(),
+                                decoded
+                                    .params
+                                    .into_iter()
+                                    .map(|param| (param.name, self.apply_label(&param.value)))
+                                    .collect(),
+                            )
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    fn apply_label(&self, token: &Token) -> String {
+        match token {
+            Token::Address(addr) => {
+                if let Some(label) = self.labels.get(addr) {
+                    format!("{}: [{:?}]", label, addr)
+                } else {
+                    format_token(token)
+                }
+            }
+            _ => format_token(token),
+        }
+    }
+
+    fn decode_cheatcode_inputs(&self, func: &Function, data: &[u8]) -> Option<Vec<String>> {
+        match func.name.as_str() {
+            "expectRevert" => foundry_utils::decode_revert(data, Some(&self.errors))
+                .ok()
+                .map(|decoded| vec![decoded]),
+            _ => None,
+        }
+    }
```

Comment:
I put this in a function by itself in case we want to do more cheatcode-specific decoding